src/util/color_ramp.js

```diff
@@ -1,28 +1,63 @@
 // @flow
 
 import {RGBAImage} from './image';
+import {isPowerOfTwo, mapValue} from './util';
+import assert from 'assert';
 
 import type {StylePropertyExpression} from '../style-spec/expression/index';
 
+export type ColorRampParams = {
+    expression: StylePropertyExpression;
+    evaluationKey: string;
+    resolution?: number;
+    image?: RGBAImage;
+    clips?: Array<Object>;
+}
+
 /**
- * Given an expression that should evaluate to a color ramp, return
- * a 256x1 px RGBA image representing that ramp expression.
+ * Given an expression that should evaluate to a color ramp,
+ * return a RGBA image representing that ramp expression.
  *
  * @private
  */
-export default function renderColorRamp(expression: StylePropertyExpression, colorRampEvaluationParameter: string): RGBAImage {
-    const colorRampData = new Uint8Array(256 * 4);
+export function renderColorRamp(params: ColorRampParams): RGBAImage {
     const evaluationGlobals = {};
-    for (let i = 0, j = 0; i < 256; i++, j += 4) {
-        evaluationGlobals[colorRampEvaluationParameter] = i / 255;
-        const pxColor = expression.evaluate((evaluationGlobals: any));
+    const width = params.resolution || 256;
+    const height = params.clips ? params.clips.length : 1;
+    const image = params.image || new RGBAImage({width, height});
+
+    assert(isPowerOfTwo(width));
+
+    const renderPixel = (stride, index, progress) => {
+        evaluationGlobals[params.evaluationKey] = progress;
+        const pxColor = params.expression.evaluate((evaluationGlobals: any));
         // the colors are being unpremultiplied because Color uses
         // premultiplied values, and the Texture class expects unpremultiplied ones
-        colorRampData[j + 0] = Math.floor(pxColor.r * 255 / pxColor.a);
-        colorRampData[j + 1] = Math.floor(pxColor.g * 255 / pxColor.a);
-        colorRampData[j + 2] = Math.floor(pxColor.b * 255 / pxColor.a);
-        colorRampData[j + 3] = Math.floor(pxColor.a * 255);
+        image.data[stride + index + 0] = Math.floor(pxColor.r * 255 / pxColor.a);
+        image.data[stride + index + 1] = Math.floor(pxColor.g * 255 / pxColor.a);
+        image.data[stride + index + 2] = Math.floor(pxColor.b * 255 / pxColor.a);
+        image.data[stride + index + 3] = Math.floor(pxColor.a * 255);
+    };
+
+    if (!params.clips) {
+        for (let i = 0, j = 0; i < width; i++, j += 4) {
+            const progress = i / (width - 1);
+
+            renderPixel(0, j, progress);
+        }
+    } else {
+        for (let clip = 0, stride = 0; clip < height; ++clip, stride += width * 4) {
+            for (let i = 0, j = 0; i < width; i++, j += 4) {
+                // Remap progress between clips
+                const progress = i / (width - 1);
+                const evaluationProgress = mapValue(progress, 0.0, 1.0,
+                    params.clips[clip].start,
+                    params.clips[clip].end);
```

Comment:
Looks like `mapValue` is only used in this one place, and with 0–1 range, so it's probably not necessary. How about rewriting as:

```js
const {start, end} = params.clips[clip];
const evaluationProgress = interpolate(start, end, progress);
```

Reply:
Yep good call, updated it here and in the shader.